aiida/orm/nodes/node.py

```diff
@@ -905,22 +909,38 @@ def _iter_all_same_nodes(self, allow_before_store=False) -> Iterator['Node']:
         """
         if not allow_before_store and not self.is_stored:
             raise exceptions.InvalidOperation('You can get the hash only after having stored the node')
+
         node_hash = self._get_hash()
 
         if not node_hash or not self._cachable:
             return iter(())
 
         builder = QueryBuilder(backend=self.backend)
-        builder.append(self.__class__, filters={'extras._aiida_hash': node_hash}, project='*', subclassing=False)
-        nodes_identical = (n[0] for n in builder.iterall())
+        builder.append(self.__class__, filters={f'extras.{self._HASH_EXTRA_KEY}': node_hash}, subclassing=False)
```

Comment:
Can't we just delegate the query builder to already filter by node.is_valid_cache, rather than doing it in python below via the getter? I imagine this can only be more efficient (one needs to be careful to the default value if not set, of course)

Reply:
This would be possible if the extra is the only thing that determines whether the node is a valid caching source. However, the method can and actually _is_ already overridden by the `ProcessNode` class. It also considers the exit code that is stored, which if it has the `invalidates_cache` set to `True` on the associated `ProcessSpec` then `is_valid_cache` also returns `False`.